### PR TITLE
Custom Domains

### PR DESCRIPTION
**Pull request by istarkov**
## Description

How it looks. (Not everything aligned or have features like in design)

https://github.com/webstudio-is/webstudio-builder/assets/5077042/349c7fae-afd8-478d-9eba-701b002f0fd4

Most API methods return `{ success: true } |  {success: false, error: string}` because in the future they need to be moved into server workflows, and do not exist or return anything.

## Steps for reproduction

Run locally, at localhost I've created a fake domain creation API, to be able to test UI.

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-builder/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

File: apps/builder/app/builder/features/topbar/collapsible-domain-section.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import {
+  Box,
+  Flex,
+  Label,
+  theme,
+  SectionTitle,
+} from "@webstudio-is/design-system";
+import React, { useState } from "react";
```
Comment (TrySound) [inline apps/builder/app/builder/features/topbar/collapsible-domain-section.tsx:8]: type ReactNode

File: packages/project-build/src/schema/styles.ts [hunk 2]
```diff
@@ -26,6 +32,6 @@ export const StylesList = z.array(StyleDecl);
 
 export type StylesList = z.infer<typeof StylesList>;
 
-export const Styles = z.map(z.string() as z.ZodType<StyleDeclKey>, StyleDecl);
+export const Styles = z.map(z.string(), StyleDecl);
 
-export type Styles = z.infer<typeof Styles>;
+export type Styles = Map<string, StyleDecl>;
```
Comment (TrySound) [inline packages/project-build/src/schema/styles.ts:37]: :(

**Review by TrySound — APPROVED** [review]

File: packages/domain/src/trpc/domain.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            // @todo: useless and must be provided by the rest/project endpoint
+            projectName: project.domain,
+            // To support preview deployments
+            // @todo: useless and must be provided by the rest/project endpoint
+            branchName: domainEnv.BRANCH_NAME,
+          }),
+        });
+
+        const text = await response.text();
+        if (!response.ok) {
```
Comment (kof) [inline packages/domain/src/trpc/domain.ts:70]:
```suggestion
        if (response.ok === false) {
```

File: packages/trpc-interface/src/context/context.server.ts [hunk 4]
```diff
@@ -23,11 +23,23 @@ type AuthorizationContext = {
   authorizeTrpc: TrpcInterfaceClient["authorize"];
 };
 
+type DomainContext = {
+  domainTrpc: TrpcInterfaceClient["domain"];
+  domainEnv: {
+    PUBLISHER_TOKEN: string;
+    PUBLISHER_ENDPOINT?: string;
```
Comment (kof) [inline packages/trpc-interface/src/context/context.server.ts:30]: I wonder if this should ever be optional

File: packages/trpc-interface/src/shared/domain.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        domainStates.set(input.domain, "pending");
+
+        return {
+          success: false,
+          error: `Domain ${input.domain} is not active`,
+        };
+      }
+
+      if (domainState === "active") {
+        setTimeout(() => {
```
Comment (kof) [inline packages/trpc-interface/src/shared/domain.ts:90]: very suspicious timeout, why?

File: packages/trpc-interface/src/shared/domain.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+          domainStates.set(input.domain, "error");
+        });
+      }
+
+      if (domainState === "pending" || domainState === "error") {
+        setTimeout(() => {
```
Comment (kof) [inline packages/trpc-interface/src/shared/domain.ts:96]: very suspicious timeout, why?

**Review by kof — APPROVED** [review]

File: packages/trpc-interface/src/context/context.server.ts (hunk 4 above)
Comment (istarkov) [reply packages/trpc-interface/src/context/context.server.ts:30]: Im gonna remove it in next iterations. And use trpc for publish

File: packages/trpc-interface/src/shared/domain.ts (hunk 6 above)
Comment (istarkov) [reply packages/trpc-interface/src/shared/domain.ts:96]: Added comment at top, its localhost/oss playground,
allows me to test UI etc

**Pull request merged**